geode-core/src/distributedTest/java/org/apache/geode/internal/cache/ClientServerTransactionDUnitTest.java

```diff
@@ -2972,6 +2974,105 @@ public Object call() throws Exception {
     });
   }
 
+  /**
+   * start 3 servers with region r1. stop client's server after distributing
+   * commit but before replying to client. ensure that a listener in the
+   * client is only invoked once
+   */
+  @Test
+  public void testFailoverAfterCommitDistributionInvokesListenerInClientOnlyOnce() {
+    Host host = Host.getHost(0);
+    VM server = host.getVM(0);
+    VM datastore1 = host.getVM(1);
+    VM datastore2 = host.getVM(2);
+    VM client = host.getVM(3);
+
+    final int port1 = createRegionsAndStartServer(server, false);
+    final int port2 = AvailablePort.getRandomAvailablePort(AvailablePort.SOCKET);
+
+    server.invoke(new CreateReplicateRegion("r1"));
+
+
+    client.invoke("create client cache", () -> {
+      disconnectFromDS();
+      System.setProperty(DistributionConfig.GEMFIRE_PREFIX + "bridge.disableShufflingOfEndpoints",
+          "true");
+      ClientCacheFactory ccf = new ClientCacheFactory();
+      ccf.addPoolServer("localhost"/* getServerHostName(Host.getHost(0)) */, port1);
+      ccf.addPoolServer("localhost", port2);
+      ccf.setPoolMinConnections(5);
+      ccf.setPoolLoadConditioningInterval(-1);
+      ccf.setPoolSubscriptionEnabled(false);
+      ccf.set(LOG_LEVEL, getDUnitLogLevel());
+      ClientCache cCache = getClientCache(ccf);
+      Region r1 =
+          cCache.createClientRegionFactory(ClientRegionShortcut.CACHING_PROXY).create("r1");
+      Region r2 =
+          cCache.createClientRegionFactory(ClientRegionShortcut.CACHING_PROXY).create("r2");
+      return null;
+    });
+
+    datastore1.invoke("create backup server", () -> {
+      CacheServer s = getCache().addCacheServer();
+      getCache().getLogger().info("SWAP:ds1");
+      s.setPort(port2);
+      s.start();
+      return null;
+    });
+    datastore1.invoke(new CreateReplicateRegion("r1"));
+    datastore2.invoke(new CreateReplicateRegion("r1"));
+
+    final TransactionId txId = client.invoke("start transaction in client", () -> {
+      ClientCache cCache = (ClientCache) getCache();
+      Region r1 = cCache.getRegion("r1");
+      r1.put("key", "value");
+      cCache.getCacheTransactionManager().begin();
+      cCache.getLogger().info("beganTX");
+      r1.destroy("key");
+      return cCache.getCacheTransactionManager().getTransactionId();
```

Comment:
I think we need to use txManager.suspend here instead. The later commit needs to resume the transaction. The reason is that we may not have the same thread on client vm to invoke these.